File: examples/latest/navigate/ios/Positioning/Positioning/PositioningExample.swift
/*
 * Copyright (C) 2019-2024 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

// CoreLocation is needed for CLLocationManagerDelegate.
import CoreLocation
import heresdk
import SwiftUI

class PositioningExample: NSObject, CLLocationManagerDelegate, LocationDelegate, LocationStatusDelegate {

    private static let defaultGeoCoordinates = GeoCoordinates(latitude: 52.520798, longitude: 13.409408)
    private static let defaultCameraDistance = 1000.0

    private var mapView: MapView
    private var locationEngine: LocationEngine
    private var locationIndicator: LocationIndicator!

    // This core location instance is needed for requesting location authorization from iOS.
    private let clLocationManager = CLLocationManager()

    init(_ mapView: MapView) {
        self.mapView = mapView

        // Create instance of location engine.
        do {
            try locationEngine = LocationEngine()
        } catch let engineInstantiationError {
            fatalError("Failed to initialize LocationEngine. Cause: \(engineInstantiationError)")
        }

        super.init()

        // Load the map scene using a map scheme to render the map with.
        mapView.mapScene.loadScene(mapScheme: MapScheme.normalDay, completion: onLoadScene)
    }

    // Completion handler for loadScene().
    private func onLoadScene(mapError: MapError?) {
        guard mapError == nil else {
            print("Error: Map scene not loaded, \(String(describing: mapError))")
            return
        }

        addMyInitialLocation()
        startLocating()
    }

    // A last known location may be already available after initialization of the location engine.
    private func addMyInitialLocation() {
        if let lastLocation = locationEngine.lastKnownLocation {
            addMyLocationToMap(myLocation: lastLocation)
        } else {
            // Fallback: It seems, this app never received a location signal so far.
            // Maybe this is the 1st time the app starts?
            var defaultLocation = Location(coordinates: PositioningExample.defaultGeoCoordinates)
            defaultLocation.time = Date()
            addMyLocationToMap(myLocation: defaultLocation)
        }
    }

    private func startLocating() {
        // Enable background updates.
        _ = locationEngine.setBackgroundLocationAllowed(allowed: true)
        _ = locationEngine.setBackgroundLocationIndicatorVisible(visible: true)

        // Set delegates and start location engine.
        locationEngine.addLocationStatusDelegate(locationStatusDelegate: self)
        locationEngine.addLocationDelegate(locationDelegate: self)

        // Make sure to add the required location permissions in the "Info.plist" file.
        // Xcode contains already "Info.plist values" in the "Build Settings" tab.
        // You can add new ones in a separate file (these will be combined with the existing values):
        // 1. In Xcode, click on your project in the Project Navigator.
        // 2. Under Targets, select the "Info" tab.
        // 3. Expand "URL Types (1)" and click on the "+" button.
        // 4. A new Info.plist file will appear in the project. You can edit now the values.

        // Start the location engine and check if permissions are missing.
        if locationEngine.start(locationAccuracy: .bestAvailable) == .missingPermissions {
            // Location permissions need to be requested from iOS. The user needs to accept
            // to be able to receive location signals.
            requestLocationAuthorization()
        }
    }

    public func requestLocationAuthorization() {
        clLocationManager.delegate = self

        if let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
           let rootViewController = windowScene.windows.first(where: { $0.isKeyWindow })?.rootViewController {

            let alert = UIAlertController(
                title: "Location access required",
                message: "This example requires location access to function correctly, please accept location access in following dialog.",
                preferredStyle: .alert
            )

            alert.addAction(UIAlertAction(title: "OK", style: .default, handler: { _ in
                // Handle OK button action.
                self.clLocationManager.requestWhenInUseAuthorization()
                alert.dismiss(animated: true, completion: nil)
            }))

            rootViewController.present(alert, animated: true, completion: nil)
        }
    }

    // Optional protocol invoked by CLLocationManagerDelegate when user permissions have been changed.
    func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        // Start location engine again and check if permissions have been granted.
        startLocating()
    }

    deinit {
        // Called when the ContentView holding an instance of this class will be deallocated.
        stopLocating()
    }

    public func stopLocating() {
        // Remove delegates and stop location engine.
        locationEngine.removeLocationDelegate(locationDelegate: self)
        locationEngine.removeLocationStatusDelegate(locationStatusDelegate: self)
        locationEngine.stop()
    }

    // Conform to LocationDelegate protocol.
    // Called by the LocationEngine when a new location signal is available.
    func onLocationUpdated(_ location: heresdk.Location) {
        updateMyLocationOnMap(myLocation: location)
        print("Location updated: \(location.coordinates)")
        print("Horizontal accuracy (m): \(String(describing: location.horizontalAccuracyInMeters))")
        print("Altitude (m): \(String(describing: location.verticalAccuracyInMeters))")
    }

    private func updateMyLocationOnMap(myLocation: Location) {
        // Update location indicator.
        locationIndicator.updateLocation(myLocation)
        // Point camera to current location.
        mapView.camera.lookAt(point: myLocation.coordinates)
    }

    // Conform to LocationStatusDelegate protocol.
    func onStatusChanged(locationEngineStatus: LocationEngineStatus) {
        print("Location engine status: \(locationEngineStatus)")
    }

    // Conform to LocationStatusDelegate protocol.
    func onFeaturesNotAvailable(features: [LocationFeature]) {
        for feature in features {
            print("Location feature not available: '%s'", String(describing: feature))
        }
    }

    private func addMyLocationToMap(myLocation: Location) {
        // Setup location indicator.
        locationIndicator = LocationIndicator()

        // Enable a halo to indicate the horizontal accuracy.
        locationIndicator.isAccuracyVisualized = true
        locationIndicator.locationIndicatorStyle = .pedestrian;
        locationIndicator.updateLocation(myLocation)
        locationIndicator.enable(for: mapView)

        // Point camera to current location.
        let distanceInMeters = MapMeasure(kind: .distance, value: PositioningExample.defaultCameraDistance)
        mapView.camera.lookAt(point: myLocation.coordinates,
                              zoom: distanceInMeters)
    }
}

File: examples/latest/navigate/flutter/positioning_app/lib/PositioningExample.dart
/*
 * Copyright (C) 2020-2025 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

import 'dart:io' show Platform, Process;

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:here_sdk/core.dart';
import 'package:here_sdk/core.engine.dart';
import 'package:here_sdk/core.errors.dart';
import 'package:here_sdk/location.dart';
import 'package:here_sdk/mapview.dart';
import 'package:permission_handler/permission_handler.dart';
import 'package:shared_preferences/shared_preferences.dart';

import 'PositioningTermsAndPrivacyHelper.dart';
import 'main.dart';

class PositioningExample extends State<MyApp>
    with WidgetsBindingObserver
    implements LocationListener, LocationStatusListener, LocationIssueListener {
  static const String _notAvailable = "--";
  static const double _spacing = 8;
  static const double _padding = 20;
  static const double _labelWidth = 100;
  static const double _labelHeight = 20;
  static const double _cameraDistanceInMeters = 400;
  static final GeoCoordinates _defaultGeoCoordinates = GeoCoordinates(52.530932, 13.384915);

  LocationEngine? _locationEngine;

  HereMapController? _hereMapController;
  LocationIndicator? _locationIndicator;
  Location? _location;
  LocationEngineStatus? _status;
  List<LocationIssueType> _issues = <LocationIssueType>[];
  late final AppLifecycleListener _appLifecycleListener;

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addObserver(this);

    _appLifecycleListener = AppLifecycleListener(
      onDetach: () =>
      // Sometimes Flutter may not reliably call dispose(),
      // therefore it is recommended to dispose the HERE SDK
      // also when the AppLifecycleListener is detached.
      // See more details: https://github.com/flutter/flutter/issues/40940
      { print('AppLifecycleListener detached.'), _disposeHERESDK() },
    );
  }

  @override
  void dispose() {
    WidgetsBinding.instance.removeObserver(this);
    _disposeHERESDK();
    super.dispose();
  }

  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
    if (state == AppLifecycleState.detached) {
      _stopLocating();
    } else if (state == AppLifecycleState.resumed) {
      if (_locationEngine != null) {
        _startLocating();
      }
    }
  }

  void _disposeHERESDK() async {
    // Free HERE SDK resources before the application shuts down.
    await SDKNativeEngine.sharedInstance?.dispose();
    SdkContext.release();
    _appLifecycleListener.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: Text('HERE SDK - Positioning Example'),
        ),
        body: Stack(
          children: [
            HereMap(onMapCreated: _onMapCreated),
            Column(
              mainAxisAlignment: MainAxisAlignment.start,
              children: [
                Row(
                  mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                  children: [
                    IgnorePointer(
                      child: Padding(
                        padding: const EdgeInsets.all(_spacing),
                        child: Column(
                          children: [
                            _buildLabelValue(
                              'Coordinates:',
                              _getCoordinatesString(),
                            ),
                            _buildLabelValue(
                              'Horz.Accuracy:',
                              _getHorizontalAccuracyString(),
                            ),
                            _buildLabelValue(
                              'Vert.Accuracy:',
                              _getVerticalAccuracyString(),
                            ),
                            _buildLabelValue(
                              'Bearing:',
                              _getBearingString(),
                            ),
                            _buildLabelValue(
                              'Bear.Accuracy:',
                              _getBearingAccuracyString(),
                            ),
                            _buildLabelValue(
                              'Speed:',
                              _getSpeedString(),
                            ),
                            _buildLabelValue(
                              'Speed.Accuracy:',
                              _getSpeedAccuracyString(),
                            ),
                            _buildLabelValue(
                              'Timestamp:',
                              _location?.time.toString() ?? _notAvailable,
                            ),
                            _buildLabelValue(
                              'SinceBoot:',
                              _getTimestampSinceBootString(),
                            ),
                            _buildLabelValue(
                              'Issues:',
                              _issues.map((issue) => issue.toString().split('.').last).join(', '),
                            ),
                            const SizedBox(height: _padding),
                            _buildLabelValue(
                              'Status:',
                              _status == null ? _notAvailable : describeEnum(_status!),
                            ),
                          ],
                        ),
                      ),
                    ),
                  ],
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }

  Future<bool> _requestPermissions() async {
    if (!await Permission.location.serviceStatus.isEnabled) {
      return false;
    }

    if (!await Permission.location.request().isGranted) {
      return false;
    }

    // All required permissions granted.
    return true;
  }

  void _onMapCreated(HereMapController hereMapController) {
    _hereMapController = hereMapController;
    hereMapController.mapScene.loadSceneForMapScheme(MapScheme.normalDay, (MapError? error) async {
      if (error != null) {
        print('Map scene not loaded. MapError: ' + error.toString());
        return;
      }

      if (Platform.isAndroid) {
        // Shows an example of how to present application terms and a privacy policy dialog as
        // required by legal requirements when using HERE Positioning.
        // See the Positioning section in our Developer Guide for more details.
        // Afterwards, Android permissions need to be checked to allow using the device's sensors.
        final termsAndPrivacyHelper = PositioningTermsAndPrivacyHelper(context);
        await termsAndPrivacyHelper.showAppTermsAndPrivacyPolicyDialogIfNeeded();
      }

      await _handlePermissions();
    });
  }

  Future<void> _handlePermissions() async {
    // Before we start the app we want to ensure that the required permissions are handled.
    if (!await _requestPermissions()) {
      await _showDialog("Error", "Cannot start app: Location service and permissions are needed for this app.");
      // Let the user set the permissions from the system settings as fallback.
      openAppSettings();
      SystemNavigator.pop();
      return;
    }

    try {
      _locationEngine = LocationEngine();
    } on InstantiationException {
      throw ("Initialization of LocationEngine failed.");
    }

    // Once permissions are granted, start locating.
    _startLocating();
  }

  void _startLocating() {
    Location? location = _locationEngine!.lastKnownLocation;

    if (location != null) {
      print("Last known location: " +
          location.coordinates.latitude.toString() +
          ", " +
          location.coordinates.longitude.toString());
    } else {
      location = Location.withCoordinates(_defaultGeoCoordinates);
      location.time = DateTime.now();
    }

    _addMyLocationToMap(location);

    // Enable background updates on iOS.
    _locationEngine!.setBackgroundLocationAllowed(true);
    _locationEngine!.setBackgroundLocationIndicatorVisible(true);

    // Set delegates and start location engine.
    _locationEngine!.addLocationListener(this);
    _locationEngine!.addLocationStatusListener(this);
    _locationEngine!.addLocationIssueListener(this);
    if (Platform.isAndroid) {
      // By calling confirmHEREPrivacyNoticeInclusion() you confirm that this app informs on
      // data collection on Android devices, which is done for this app via PositioningTermsAndPrivacyHelper,
      // which shows a possible example for this.
      _locationEngine!.confirmHEREPrivacyNoticeInclusion();
    }
    _locationEngine!.startWithLocationAccuracy(LocationAccuracy.bestAvailable);
  }

  void _stopLocating() {
    _locationEngine!.removeLocationIssueListener(this);
    _locationEngine!.removeLocationStatusListener(this);
    _locationEngine!.removeLocationListener(this);
    _locationEngine!.stop();
  }

  void _addMyLocationToMap(Location myLocation) {
    if (_locationIndicator != null) {
      return;
    }
    // Set-up location indicator.
    _locationIndicator = LocationIndicator();
    // Enable a halo to indicate the horizontal accuracy.
    _locationIndicator!.isAccuracyVisualized = true;
    _locationIndicator!.locationIndicatorStyle = LocationIndicatorIndicatorStyle.pedestrian;
    _locationIndicator!.updateLocation(myLocation);
    _locationIndicator!.enable(_hereMapController!);

    // Point camera at given location.
    MapMeasure mapMeasureZoom = MapMeasure(MapMeasureKind.distanceInMeters, _cameraDistanceInMeters);
    _hereMapController!.camera.lookAtPointWithMeasure(
      myLocation.coordinates,
      mapMeasureZoom,
    );

    // Update state's location.
    setState(() {
      _location = myLocation;
    });
  }

  void _updateMyLocationOnMap(Location myLocation) {
    if (_locationIndicator == null) {
      return;
    }

    // Update location indicator's location.
    _locationIndicator!.updateLocation(myLocation);
    // Point camera at given location.
    _hereMapController!.camera.lookAtPoint(myLocation.coordinates);
    // Update state's location.
    setState(() {
      _location = myLocation;
    });
  }

  @override
  onFeaturesNotAvailable(List<LocationFeature> features) {
    for (var feature in features) {
      print("Feature not available: " + feature.toString());
    }
  }

  @override
  void onLocationUpdated(Location location) {
    print("Location update: " +
        location.coordinates.latitude.toString() +
        ", " +
        location.coordinates.longitude.toString());
    _updateMyLocationOnMap(location);
  }

  @override
  void onStatusChanged(LocationEngineStatus locationEngineStatus) {
    setState(() {
      _status = locationEngineStatus;
    });
  }

  @override
  void onLocationIssueChanged(List<LocationIssueType> issues) {
    setState(() => _issues = issues);
  }

  Widget _buildLabelValue(String text, String value) {
    return Row(
      children: <Widget>[
        _createLabel(text),
        const SizedBox(width: _padding),
        _createValue(value),
      ],
    );
  }

  Widget _createLabel(String text) {
    return SizedBox(
      width: _labelWidth,
      height: _labelHeight,
      child: Text(
        text,
        style: TextStyle(fontWeight: FontWeight.bold),
      ),
    );
  }

  Widget _createValue(String text) {
    return SizedBox(height: _labelHeight, child: Text(text));
  }

  // A helper method to show a dialog.
  Future<void> _showDialog(String title, String message) async {
    return showDialog<void>(
      context: context,
      barrierDismissible: false,
      builder: (BuildContext context) {
        return AlertDialog(
          title: Text(title),
          content: SingleChildScrollView(
            child: ListBody(
              children: <Widget>[
                Text(message),
              ],
            ),
          ),
          actions: <Widget>[
            TextButton(
              child: Text('OK'),
              onPressed: () {
                Navigator.of(context).pop();
              },
            ),
          ],
        );
      },
    );
  }

  String _getCoordinatesString() {
    if (_location == null) {
      return _notAvailable;
    }
    return _location!.coordinates.latitude.toStringAsFixed(6) +
        ', ' +
        _location!.coordinates.longitude.toStringAsFixed(6);
  }

  String _getHorizontalAccuracyString() {
    return '${_location?.horizontalAccuracyInMeters?.toStringAsFixed(1) ?? _notAvailable}' + ' m';
  }

  String _getVerticalAccuracyString() {
    return '${_location?.verticalAccuracyInMeters?.toStringAsFixed(1) ?? _notAvailable}' + ' m';
  }

  String _getBearingString() {
    return '${_location?.bearingInDegrees?.toStringAsFixed(1) ?? _notAvailable}' + ' °';
  }

  String _getBearingAccuracyString() {
    return '${_location?.bearingAccuracyInDegrees?.toStringAsFixed(1) ?? _notAvailable}' + ' °';
  }

  String _getSpeedString() {
    return '${_location?.speedInMetersPerSecond?.toStringAsFixed(1) ?? _notAvailable}' + ' m/s';
  }

  String _getSpeedAccuracyString() {
    return '${_location?.speedAccuracyInMetersPerSecond?.toStringAsFixed(1) ?? _notAvailable}' + ' m/s';
  }

  String _getTimestampSinceBootString() {
    return '${_location?.timestampSinceBoot?.toString() ?? _notAvailable}' + ' ms';
  }
}
